### PR TITLE
Fix: Viz not working with proxy URLs

### LINKED ISSUE #916: Kedro viz with proxy does not show additional information
## Description
Additional information is available only for the first couple of seconds when you are open it, and then it just disappears. Look at the picture below. I hided some information. What I am saying about is inside red rectangle.
Screenshot 2022-06-03 at 16 39 01

## Context
We have a remote cluster setup for working where we have jupyter hub installed and all interactions with this remote server are through this hub. So, to get kedro viz there, we had to make it work via this hub by using proxies. I am not sure about all technical details, but I can share with you my conversation with our support guy who was investigating this problem a bit:

> Long story short -- this is an issue with how kedro-viz frontend treats working with the proxy server
> Static javascript that is already built-in in the package is not configured to dynamically modify its paths when served through the proxy server (which is the case for us). On the start of kedro viz, we load an html which has some relative paths, causing all the content to prepend /user/<username>  base path. But all the subsequent requests are done from the loaded js files that don't have  /user/<username> base path prepended as they should. After lots of digging I found out this: https://create-react-app.dev/docs/proxying-api-requests-in-development/#configuring-the-proxy-manually. Which I confirmed to myself by finding this: https://github.com/mlflow/mlflow/blob/master/mlflow/server/js/src/setupProxy.js. As you can see, this file being present in their repo is most likely (99%) the reason why mlflow does not have such an issue.

I can confirm that we start MLFlow the same way as Kedro Viz, and it does not have such problems. But still, I cannot be sure that the reason is really in these configurations.

## Steps to Reproduce
I am afraid I cannot properly describe all steps to reproduce this issue, as I have lack of technical knoledges of how jupyterhub works with proxies. Please let me know if you have any questions that I can address to our infrastructure team to help you understand the source of the problem.

## Expected Result
Information should be shown

## Actual Result
Information does not be shown

## Your Environment

* Web browser system and version: does not matter
* Operating system and version: CentOS Linux 7
* Output of jupyter --version command: 
```
  IPython          : 7.29.0
  ipykernel        : 6.5.0
  ipywidgets       : 7.6.5
  jupyter_client   : 7.0.6
  jupyter_core     : 4.9.1
  jupyter_server   : 1.11.2
  jupyterlab       : 3.2.2
  nbclient         : 0.5.5
  nbconvert        : 6.2.0
  nbformat         : 5.1.3
  notebook         : 6.4.5
  qtconsole        : not installed
  traitlets        : 5.1.1
```

## Checklist
- [x] Include labels so that we can categorise your issue

### PR DESCRIPTION
**Pull request by tynandebold**
## Description

Fixes #916

## Development notes

If a user runs Viz with a proxied URL, like you may try to do as described in [this issue](https://github.com/kedro-org/kedro-viz/issues/916), the application would load without critical information, and sometimes no information at all. The reason being is that the URL wouldn't pass our [hard-coded route match](https://github.com/kedro-org/kedro-viz/blob/main/src/components/wrapper/wrapper.js#L60).

In this PR, we'll now allow a dynamic pathname to be recognized and matched, thus allowing the flowchart or experiment-tracking pages to render correctly.

Lastly, we'll also sanitize the URL and remove any duplicative `experiment-tracking` strings from appearing in the pathname. We need to guard against this because, again, if a user has a proxy URL or a base URL that isn't `/`, there will be conflicts. The use case: a user runs Viz with [jupyter-sever-proxy](https://jupyter-server-proxy.readthedocs.io/en/latest/index.html), which adds `/kedro_viz/` to the pathname. So Viz loads, then we navigate to the experiment-tracking page, which does load correctly. Yet if the page is refreshed, we'd receive an incorrect result. If we didn't remove `experiment-tracking` from the pathname, we'd get something like `/kedro_viz/experiment-trackingexperiment-tracking`, and that's not desirable.

## QA notes

Run Viz via `jupyter-sever-proxy` by doing the following:

- Checkout the branch
- `cd` in the `demo-project` folder
- `pip install jupyter-sever-proxy`
- Run `kedro viz`
- Click the `Kedro Viz` icon that appears

All functionality within Viz should function as it should.

## Checklist

- [x] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added new entries to the `RELEASE.md` file
- [ ] Added tests to cover my changes

<a href="https://gitpod.io/#https://github.com/kedro-org/kedro-viz/pull/941"></a>

### CODE REVIEW

File: src/components/icons/logo.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import React from 'react';
 
 const LogoIcon = ({ className }) => (
-  <svg className={className} viewBox="24 25 32 30">
+  <svg viewBox="24 25 32 30">
```
Comment (AntonyMilneQB) [inline src/components/icons/logo.js:4]:
```suggestion
  <svg className={className} viewBox="24 25 32 30">
```
This was just me trying to do something, wasn't meant to stay as a change.

**Review by AntonyMilneQB — COMMENTED** [review]
This is great! I will test it out later today.

Let's leave out all the Python stuff and icon for now since it shouldn't be needed just to fix the proxy bug. It needs a bit more thought (the stuff I did was just a quick POC) and would be done as part of a modified https://github.com/kedro-org/kedro-viz/issues/910.

File: package/setup.py [hunk 2]
```diff
@@ -50,6 +50,10 @@
     package_data={"kedro_viz": list(files)},
     zip_safe=False,
     entry_points={
+        "jupyter_serverproxy_servers": [
+            # name = packagename:function_name
+            "kedro_viz = kedro_viz.launchers.jupyter:jupyter_server_proxy",
+        ],
```
Comment (tynandebold) [inline package/setup.py:56]: @AntonyMilneQB should I remove this, too?

File: src/components/icons/logo.js (hunk 1 above)
Comment (tynandebold) [reply src/components/icons/logo.js]: Reverted.

File: package/setup.py (hunk 2 above)
Comment (tynandebold) [reply package/setup.py:56]: I've removed it.

File: src/apollo/config.js [hunk 3]
```diff
@@ -7,24 +7,27 @@ import {
...
-    : window.location.host;
+const { host, pathname, protocol } = window.location;
+const sanitizedPathname = replaceMatches(pathname, {
+  'experiment-tracking': '',
+});
+
+const wsHost = process.env.NODE_ENV === 'development' ? 'localhost:4142' : host;
 
-const wsProtocol = window.location.protocol === 'https:' ? 'wss' : 'ws';
+const wsProtocol = protocol === 'https:' ? 'wss' : 'ws';
```
Comment (AntonyMilneQB) [inline src/apollo/config.js:19]:
```suggestion
const wsProtocol = protocol === 'https:' ? 'wss:' : 'ws:';
```
Just to be consistent about what "protcol" refers to.

File: src/components/global-toolbar/global-toolbar.js [hunk 4]
```diff
@@ -32,7 +38,7 @@ export const GlobalToolbar = ({
             disabled={false}
             icon={LogoIcon}
           />
-          <NavLink exact to={{ pathname: '/' }}>
+          <NavLink exact to={{ pathname: `${sanitizedPathname}` }}>
```
Comment (AntonyMilneQB) [inline src/components/global-toolbar/global-toolbar.js:41]:
```suggestion
          <NavLink exact to={{ pathname: {sanitizedPathname} }}>
```
If I got the syntax right...

**Review by AntonyMilneQB — COMMENTED** [review]
This is great!! But... it doesn't actually seem to work for me 🤔 I'm not sure why because when I check the `window.location` it seems like it should. Probably I've just not set it up correctly locally. Let me show you on Wednesday and hopefully we can figure it out.

I'm also wondering:
* should we make this even more generic to include `search` and `hash` as well as `pathname`? I think this actually might be needed in real scenarios with jupyter server proxy, not just a theoretical possibility. Seems like it's possible to pass a `location` object to these tags rather than needing to reconstruct the `host + pathname + search + hash` string
* I don't quite get where the `/` should be still 😅 I'm sure what you've done is right but I need to think through it again with a fresh head to understand

**Comment by tynandebold** [comment]
> This is great!! But... it doesn't actually seem to work for me 🤔 I'm not sure why because when I check the `window.location` it seems like it should. Probably I've just not set it up correctly locally. Let me show you on Wednesday and hopefully we can figure it out.
> 
> I'm also wondering:
> 
> * should we make this even more generic to include `search` and `hash` as well as `pathname`? I think this actually might be needed in real scenarios with jupyter server proxy, not just a theoretical possibility. Seems like it's possible to pass a `location` object to these tags rather than needing to reconstruct the `host + pathname + search + hash` string
> * I don't quite get where the `/` should be still 😅 I'm sure what you've done is right but I need to think through it again with a fresh head to understand

Oh no, ok, let's sync up today.

**Review by AntonyMilneQB — APPROVED** [review]
Awesome stuff. Super excited to get this fixed and what it means we can hopefully do next 💥 

It would be nice if a future iteration also passed the `search` and `hash` (looks like [`Link` at least could handle this easily](https://v5.reactrouter.com/web/api/Link/to-object))

**Pull request merged**